docs: add replayable ai agents guide (#4950)
neondatabase/website: content/guides/replayable-ai-agents.md

- Execution checkpoints
- Suspend and resume flows
- Replayable runs and state restoration
- Time-travel debugging

The key idea is simple: **pair your agent's execution state with durable database state**. Every time the agent reaches a critical juncture, take a Neon snapshot and serialize the agent's memory. If a failure occurs, you can restore the database and resume the agent from the same logical point.

## Architecture overview

Making an agent replayable requires capturing two critical pieces of state at the moment of interruption:

1. **Agent state**: The conversation history, tool calls, and execution graph. The OpenAI Agents SDK provides this via the [`RunState`](https://openai.github.io/openai-agents-python/ref/run_state/) object, which can be serialized to JSON and resumed after a [human-in-the-loop](https://openai.github.io/openai-agents-python/human_in_the_loop/) interruption.
2. **Database state**: The schema and data at the moment of the pause. Neon handles this with copy-on-write **snapshots**, which capture large databases instantly without heavy duplication.

### What are Neon Snapshots?

> A [Snapshot](/docs/reference/glossary#snapshot) in Neon is an immutable, point-in-time backup of your database branch's schema and data. While standard database branching is great for development - creating an isolated environment with a new connection string, snapshots are built for versioning and time-travel recovery.

Reviewer: ```suggestion
A [Snapshot](/docs/reference/glossary#snapshot) in Neon is an immutable, point-in-time backup of your database branch's schema and data. While standard database branching is great for development and creating an isolated environment with a new connection string, snapshots are built for versioning and time-travel recovery.
```

- **Stable connections:** By using `target_branch_id` and `finalize_restore: True` in the Neon API, the compute endpoint for your database moves to the restored state. Your application connection string does not have to change.
- **Auditable AI:** By serializing the `RunState` alongside database checkpoints, you maintain an audit log of what an agent saw, planned, and executed.

With this pattern, your AI agents are no longer black boxes making irreversible changes. They are fully versioned, durable, and easily debuggable systems. The recovery flow above shows the immediate case; the same idea also works for older runs that you need to replay without touching production.

## Replay historical runs safely with branches

You demonstrated an immediate rollback by restoring the database in place using `finalize_restore: True`. But what happens if you discover a bad agent decision **10 days later**?

Imagine your agent processed a massive financial reconciliation workflow a week ago. Today, you realize the agent's prompt had a subtle hallucination, and it categorized a batch of transactions incorrectly. You want to replay that exact execution from 10 days ago and fix it.

> You cannot restore your production database in-place to 10 days ago - you would wipe out a week of real user activity!

Reviewer: ```suggestion
You cannot restore your production database in-place to 10 days ago. You would wipe out a week of real user activity!
```

2. **Load the historical agent:** You retrieve the agent's 10-day-old `RunState` JSON from your object store or metadata database and load it into memory.
3. **Inject the new connection string:** You pass the new, isolated branch's `DATABASE_URL` into the agent's environment.
4. **Replay and correct:** You reject the historical hallucination, provide the corrected prompt, and let the agent re-run the reconciliation in the isolated branch.
5. **Extract or merge:** Once the agent successfully completes the task in the isolated branch, write a script to extract the corrected transaction data and safely apply it to today's production database.

By combining serialized agent memory with copy-on-write database branches, you can treat your agent's actions like versioned changes: check out an old state, create a branch, fix the bug, and merge the results back deliberately.

## Scaling this workflow to production

The examples in this guide show the minimal components required for replayability: a Postgres database, a sample `users` table, and a local `checkpoint.json` file.

> In practice, AI systems run inside a robust **agent harness** using an asynchronous, event-driven architecture. The core primitive - pairing an agent state with a Neon snapshot remains the same, but the surrounding infrastructure ensures the system is durable and maintainable:

Reviewer: ```suggestion
In practice, AI systems run inside a robust **agent harness** using an asynchronous, event-driven architecture. The core primitive that pairs an agent state with a Neon snapshot remains the same, but the surrounding infrastructure ensures the system is durable and maintainable:
```
